File: platform/polyglot/github_api/py/github_api/routes/branches.py
"""
Branch routes for the FastAPI server.

Maps HTTP endpoints to BranchesClient methods.
"""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter, Request

__all__ = ["router"]

router = APIRouter(tags=["branches"])


def _get_branches_client(request: Request):
    """Get the BranchesClient from app state."""
    from github_api.sdk.branches import BranchesClient
    return BranchesClient(request.app.state.github_client)


@router.get("/repos/{owner}/{repo}/branches")
async def list_branches(request: Request, owner: str, repo: str) -> dict[str, Any]:
    """List branches for a repository."""
    client = _get_branches_client(request)
    params = dict(request.query_params)
    return await client.list(owner, repo, params=params or None)


@router.get("/repos/{owner}/{repo}/branches/{branch}")
async def get_branch(
    request: Request, owner: str, repo: str, branch: str
) -> dict[str, Any]:
    """Get a specific branch."""
    client = _get_branches_client(request)
    return await client.get(owner, repo, branch)


@router.get("/repos/{owner}/{repo}/branches/{branch}/protection")
async def get_branch_protection(
    request: Request, owner: str, repo: str, branch: str
) -> dict[str, Any]:
    """Get branch protection rules."""
    client = _get_branches_client(request)
    return await client.get_protection(owner, repo, branch)


@router.put("/repos/{owner}/{repo}/branches/{branch}/protection")
async def update_branch_protection(
    request: Request, owner: str, repo: str, branch: str
) -> dict[str, Any]:
    """Update branch protection rules."""
    client = _get_branches_client(request)
    body = await request.json()
    return await client.update_protection(owner, repo, branch, body)


@router.delete("/repos/{owner}/{repo}/branches/{branch}/protection")
async def remove_branch_protection(
    request: Request, owner: str, repo: str, branch: str
) -> dict[str, Any]:
    """Remove branch protection."""
    client = _get_branches_client(request)
    return await client.remove_protection(owner, repo, branch)


@router.post("/repos/{owner}/{repo}/branches/{branch}/rename")
async def rename_branch(
    request: Request, owner: str, repo: str, branch: str
) -> dict[str, Any]:
    """Rename a branch."""
    client = _get_branches_client(request)
    body = await request.json()
    return await client.rename(owner, repo, branch, body["new_name"])


@router.post("/repos/{owner}/{repo}/merges")
async def merge_branches(request: Request, owner: str, repo: str) -> dict[str, Any]:
    """Perform a branch merge."""
    client = _get_branches_client(request)
    body = await request.json()
    return await client.merge(owner, repo, body)


@router.get("/repos/{owner}/{repo}/compare/{base_head}")
async def compare_branches(
    request: Request, owner: str, repo: str, base_head: str
) -> dict[str, Any]:
    """Compare two branches/commits/tags.

    The base_head parameter should be in the format 'base...head'.
    """
    client = _get_branches_client(request)
    parts = base_head.split("...", 1)
    if len(parts) != 2:
        from github_api.sdk.errors import ValidationError
        raise ValidationError(
            f"Invalid comparison format: {base_head!r}. Expected 'base...head'."
        )
    return await client.compare(owner, repo, parts[0], parts[1])

File: platform/polyglot/github_api/mjs/src/routes/branches.mjs
/**
 * Branch routes.
 * Exposes GitHub Branch API operations as REST endpoints.
 * @module routes/branches
 */

/**
 * Fastify plugin that registers branch routes.
 * @param {import('fastify').FastifyInstance} fastify
 * @param {Object} opts
 * @param {import('../sdk/branches/client.mjs').BranchesClient} opts.branches - Branches SDK client
 */
export default async function branchRoutes(fastify, opts) {
  const { branches } = opts;

  /**
   * GET /repos/:owner/:repo/branches - List branches.
   */
  fastify.get('/repos/:owner/:repo/branches', async (request, reply) => {
    const { owner, repo } = request.params;
    const result = await branches.list(owner, repo, request.query);
    return reply.send(result);
  });

  /**
   * GET /repos/:owner/:repo/branches/:branch - Get a branch.
   */
  fastify.get(
    '/repos/:owner/:repo/branches/:branch',
    async (request, reply) => {
      const { owner, repo, branch } = request.params;
      const result = await branches.get(owner, repo, branch);
      return reply.send(result);
    },
  );

  /**
   * GET /repos/:owner/:repo/branches/:branch/protection - Get branch protection.
   */
  fastify.get(
    '/repos/:owner/:repo/branches/:branch/protection',
    async (request, reply) => {
      const { owner, repo, branch } = request.params;
      const result = await branches.getProtection(owner, repo, branch);
      return reply.send(result);
    },
  );

  /**
   * PUT /repos/:owner/:repo/branches/:branch/protection - Update branch protection.
   */
  fastify.put(
    '/repos/:owner/:repo/branches/:branch/protection',
    async (request, reply) => {
      const { owner, repo, branch } = request.params;
      const result = await branches.updateProtection(
        owner,
        repo,
        branch,
        request.body,
      );
      return reply.send(result);
    },
  );

  /**
   * DELETE /repos/:owner/:repo/branches/:branch/protection - Remove branch protection.
   */
  fastify.delete(
    '/repos/:owner/:repo/branches/:branch/protection',
    async (request, reply) => {
      const { owner, repo, branch } = request.params;
      await branches.removeProtection(owner, repo, branch);
      return reply.status(204).send();
    },
  );

  /**
   * POST /repos/:owner/:repo/branches/:branch/rename - Rename a branch.
   */
  fastify.post(
    '/repos/:owner/:repo/branches/:branch/rename',
    async (request, reply) => {
      const { owner, repo, branch } = request.params;
      const { new_name } = request.body;
      const result = await branches.rename(owner, repo, branch, new_name);
      return reply.send(result);
    },
  );

  /**
   * POST /repos/:owner/:repo/merges - Merge branches.
   */
  fastify.post('/repos/:owner/:repo/merges', async (request, reply) => {
    const { owner, repo } = request.params;
    const { base, head, commit_message } = request.body;
    const result = await branches.merge(owner, repo, base, head, {
      commit_message,
    });
    return reply.status(201).send(result);
  });

  /**
   * GET /repos/:owner/:repo/compare/:base...:head - Compare two refs.
   */
  fastify.get(
    '/repos/:owner/:repo/compare/:base...:head',
    async (request, reply) => {
      const { owner, repo, base, head } = request.params;
      const result = await branches.compare(owner, repo, base, head);
      return reply.send(result);
    },
  );

  return Promise.resolve();
}
